WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/FIle List/ProductDownloadListViewModel.swift

```diff
@@ -0,0 +1,132 @@
+import Foundation
+import Yosemite
+
+/// Provides data needed for downloadable files settings.
+///
+protocol ProductDownloadListViewModelOutput {
+    var downloadableFiles: [ProductDownloadDragAndDrop] { get }
+    var downloadLimit: Int64 { get }
+    var downloadExpiry: Int64 { get }
+
+    // Convenience Methods
+    @discardableResult
+    func remove(at index: Int) -> ProductDownloadDragAndDrop?
+    func item(at index: Int) -> ProductDownloadDragAndDrop?
+    func insert(_ newElement: ProductDownloadDragAndDrop, at index: Int)
+    func count() -> Int
+}
+
+/// Handles actions related to the downloadable files settings data.
+///
+protocol ProductDownloadListActionHandler {
+    // Tap actions
+    func didTapDownloadableFileFromRow(_ indexPath: IndexPath)
+
+    // Input field actions
+    func handleDownloadsChange(_ downloads: [ProductDownload])
+    func handleDownloadLimitChange(_ downloadLimit: Int64)
+    func handleDownloadExpiryChange(_ downloadExpiry: Int64)
+
+    // Navigation actions
+    func completeUpdating(onCompletion: ProductDownloadListViewController.Completion)
+    func hasUnsavedChanges() -> Bool
+}
+
+/// Error cases that could occur in product downloadable files settings.
+///
+enum ProductDownloadListError: Error {
+    case emptyFileName
+    case emptyUrl
+}
+
+/// Provides view data for downloadable files settings, and handles init/UI/navigation actions needed in product downloadable files settings.
+///
+final class ProductDownloadListViewModel: ProductDownloadListViewModelOutput {
+    private let product: ProductFormDataModel
+
+    // Editable data
+    //
+    private(set) var downloadableFiles = [ProductDownloadDragAndDrop]()
+    private(set) var downloadLimit: Int64
+    private(set) var downloadExpiry: Int64
+
+    init(product: ProductFormDataModel) {
+        self.product = product
+
+        downloadableFiles = product.downloadableFiles.map { ProductDownloadDragAndDrop(downloadableFile: $0) }
+        downloadLimit = product.downloadLimit
+        downloadExpiry = product.downloadExpiry
+    }
+
+    // MARK: - ProductDownloadListDataSource Methods
+    //
+    func remove(at index: Int) -> ProductDownloadDragAndDrop? {
+        return index < downloadableFiles.count ? downloadableFiles.remove(at: index) : nil
+    }
+
+    func item(at index: Int) -> ProductDownloadDragAndDrop? {
+        return index < downloadableFiles.count ? downloadableFiles[index] : nil
```

Comment:
You can use this shorter version:
```suggestion
        return downloadableFiles[safe: index]
```

Reply:
Perfect